winit-web/src/web_sys/event.rs

```diff
@@ -24,59 +24,52 @@ bitflags::bitflags! {
-            _ => MouseButton::Other(value.bits()),
-        }
-    }
+pub fn mouse_buttons(event: &MouseEvent) -> ButtonsState {
+    ButtonsState::from_bits_retain(event.buttons())
 }
 
-impl From<MouseButton> for ButtonsState {
-    fn from(value: MouseButton) -> Self {
-        match value {
-            MouseButton::Left => ButtonsState::LEFT,
-            MouseButton::Right => ButtonsState::RIGHT,
-            MouseButton::Middle => ButtonsState::MIDDLE,
-            MouseButton::Back => ButtonsState::BACK,
-            MouseButton::Forward => ButtonsState::FORWARD,
-            MouseButton::Other(value) => ButtonsState::from_bits_retain(value),
-        }
+// https://www.w3.org/TR/pointerevents3/#the-button-property
+#[derive(Clone, Copy)]
+pub struct MouseButton(u16);
+
+impl MouseButton {
+    // raw code
+    pub fn raw(self) -> u16 {
+        self.0
     }
-}
 
-pub fn mouse_buttons(event: &MouseEvent) -> ButtonsState {
-    ButtonsState::from_bits_retain(event.buttons())
+    // Map known buttons to [`ButtonsState`], unknown to an empty state
+    pub fn state(self) -> ButtonsState {
+        match self.0 {
+            0 => ButtonsState::LEFT,
+            1 => ButtonsState::MIDDLE,
+            2 => ButtonsState::RIGHT,
+            3 => ButtonsState::BACK,
+            4 => ButtonsState::FORWARD,
+            _ => ButtonsState::empty(),
+        }
+    }
 }
 
 pub fn mouse_button(event: &MouseEvent) -> Option<MouseButton> {
-    // https://www.w3.org/TR/pointerevents3/#the-button-property
-    match event.button() {
-        -1 => None,
-        0 => Some(MouseButton::Left),
-        1 => Some(MouseButton::Middle),
-        2 => Some(MouseButton::Right),
-        3 => Some(MouseButton::Back),
-        4 => Some(MouseButton::Forward),
-        i => {
-            Some(MouseButton::Other(i.try_into().expect("unexpected negative mouse button value")))
-        },
-    }
+    // NOTE: any negative value gets mapped to `None`, though only -1 is documented
+    event.button().try_into().ok().map(MouseButton)
 }
 
-pub fn mouse_button_to_id(button: MouseButton) -> u16 {
-    match button {
-        MouseButton::Left => 0,
-        MouseButton::Right => 1,
-        MouseButton::Middle => 2,
-        MouseButton::Back => 3,
-        MouseButton::Forward => 4,
-        MouseButton::Other(value) => value,
+impl From<MouseButton> for winit_core::event::ButtonSource {
+    fn from(button: MouseButton) -> Self {
+        use winit_core::event::{ButtonSource, MouseButton as MB};
+
+        match button.0 {
+            0 => MB::LEFT.into(),
+            1 => MB::MIDDLE.into(),
+            2 => MB::RIGHT.into(),
+            3 => MB::BACK.into(),
+            4 => MB::FORWARD.into(),
+            // Codes above 4 are not observed on Firefox or Chromium. 5 is defined as an eraser,
+            // which is not a mouse button. No other codes are defined.
+            i => ButtonSource::Unknown(i),
+        }
```

Comment:
The "eraser" button could map to `ButtonSource::Unknown` on web, except that this code path *shouldn't* be used for non-mouse input. In my testing (using web forms not winit) I never saw any other codes output. So I don't *think* web platforms can produce any `Unknown` output codes in practice, though that could conceivably change in the future.

Reply:
It's part of "the button property" spec:

https://www.w3.org/TR/pointerevents3/#the-button-property